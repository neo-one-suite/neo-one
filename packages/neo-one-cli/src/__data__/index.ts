export * from './getClients';
export * from './getContracts';
export * from './buildTests';
export * from './generateTests';
