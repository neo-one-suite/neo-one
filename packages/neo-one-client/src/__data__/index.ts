export * from './keys';
export * from './transactions';
