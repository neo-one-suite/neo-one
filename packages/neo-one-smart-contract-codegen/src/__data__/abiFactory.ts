import { ABIDefault, ABIReturn, ContractABIClient } from '@neo-one/client-common';
import _ from 'lodash';

// tslint:disable-next-line:no-any
const getCombinations = (list: ReadonlyArray<ReadonlyArray<any>>) => {
  const floorsIn = list.reduceRight<ReadonlyArray<number>>((acc, right) => [right.length * acc[0]].concat(acc), [1]);
  const max = floorsIn[0];
  const floors = floorsIn.slice(1);

  const iteratorFuncs = list.map((item, index) =>
    index === list.length - 1
      ? (value: number) => item[value % item.length]
      : (value: number) => item[Math.floor(value / floors[index]) % item.length],
  );

  return _.range(max).map((_unused, index) => iteratorFuncs.map((func) => func(index)));
};

const abiFunctionOptions: ReadonlyArray<{ readonly [key in string]?: boolean }> = [
  { constant: true },
  { send: true },
  { sendUnsafe: true },
  { sendUnsafe: true, receive: true },
  { receive: true },
  { claim: true },
  { refundAssets: true },
  { completeSend: true },
  {},
];

const abiParameterOptions: ReadonlyArray<{ readonly [key in string]?: boolean | ABIDefault }> = [
  { default: { type: 'sender' } },
  { rest: true },
  { default: { type: 'sender' }, rest: true },
  {},
];

const abiReturns: ReadonlyArray<ABIReturn> = [
  {
    type: 'ForwardValue',
  },
  {
    type: 'Boolean',
    optional: true,
  },
  {
    type: 'Boolean',
    forwardedValue: true,
  },
  {
    type: 'Integer',
    decimals: 4,
  },
  {
    type: 'Array',
    value: {
      type: 'Boolean',
    },
  },
  {
    type: 'Map',
    key: { type: 'Integer', decimals: 0 },
    value: { type: 'Boolean' },
  },
  {
    type: 'Object',
    properties: {
      property: { type: 'Boolean' },
    },
  },
];

const boolReturnType: ABIReturn = { type: 'Boolean' };

const createBaseABIFunctions = (): ReadonlyArray<ContractABIClient> => {
  const combinations = getCombinations([abiReturns, abiFunctionOptions]);

  return combinations.map(([abiReturn, abiFuncProps]) => {
    const funcPropNames = Object.keys(abiFuncProps);
    const name = `${abiReturn.type}${`${funcPropNames.length === 0 ? '' : '_'}${funcPropNames.join('_')}`}`;

    return {
      methods: [
        {
          name,
          returnType: abiReturn,
          ...abiFuncProps,
        },
      ],
    };
  });
};

const createBaseABIEvents = (): ReadonlyArray<ContractABIClient> => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      methods: [
        {
          name: 'foo',
          returnType: boolReturnType,
        },
      ],
      events: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
        },
      ],
    };
  });
};

const createBaseABIFunctionParameters = (): ReadonlyArray<ContractABIClient> => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      methods: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
          returnType: boolReturnType,
        },
      ],
    };
  });
};

export const abiFactory = {
  createBaseABIFunctions,
  createBaseABIFunctionParameters,
  createBaseABIEvents,
};
