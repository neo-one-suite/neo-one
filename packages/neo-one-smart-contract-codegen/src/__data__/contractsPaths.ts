import { ContractPaths } from '../type';

const sourceMap = {
  version: 0,
  sources: [],
  names: [],
  mappings: 'sourcemap',
  file: 'file',
};

export const contractsPaths: ReadonlyArray<ContractPaths> = [
  {
    name: 'Token',
    contractPath: '/foo/bar/one/contracts/Token.ts',
    typesPath: '/foo/bar/one/generated/Token/types.ts',
    addresses: ['AV6PhAq1FHBeCMsNUPMWzpUFxvXRFwHoLn'],
    sourceMap,
  },
  {
    name: 'ICO',
    contractPath: '/foo/bar/one/contracts/ICO.ts',
    typesPath: '/foo/bar/one/generated/ICO/types.ts',
    addresses: ['ATfAohZQ5iuGDXWjcHxLAj8WXwQchGLXhi'],
    sourceMap,
  },
];
