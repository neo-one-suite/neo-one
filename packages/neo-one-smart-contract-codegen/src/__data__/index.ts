export * from './nep17';
export * from './testUtils';
export * from './abiFactory';
