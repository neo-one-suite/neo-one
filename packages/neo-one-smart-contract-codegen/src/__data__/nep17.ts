import { ContractABIClient } from '@neo-one/client-common';

const abi = (decimals: number): ContractABIClient => ({
  hash: '',
  methods: [
    {
      name: 'name',
      constant: true,
      parameters: [],
      returnType: { type: 'String' },
      offset: 0,
    },
    {
      name: 'symbol',
      constant: true,
      parameters: [],
      returnType: { type: 'String' },
      offset: 0,
    },
    {
      name: 'decimals',
      constant: true,
      parameters: [],
      returnType: { type: 'Integer', decimals: 0 },
      offset: 0,
    },
    {
      name: 'totalSupply',
      constant: true,
      parameters: [],
      returnType: { type: 'Integer', decimals: 8 },
      offset: 0,
    },
    {
      name: 'transfer',
      parameters: [
        {
          type: 'Address',
          name: 'from',
        },
        {
          type: 'Address',
          name: 'to',
        },
        {
          type: 'Integer',
          name: 'value',
          decimals: 8,
        },
        {
          type: 'ForwardValue',
          name: 'args',
          rest: true,
        },
      ],
      returnType: { type: 'Boolean' },
      offset: 0,
    },
    {
      name: 'balanceOf',
      constant: true,
      parameters: [
        {
          type: 'Address',
          name: 'account',
        },
      ],
      returnType: { type: 'Integer', decimals },
      offset: 0,
    },
    {
      name: 'forward',
      parameters: [
        { name: 'address', type: 'Address' },
        { name: 'args', type: 'ForwardValue', rest: true },
      ],
      returnType: { type: 'ForwardValue' },
      offset: 0,
    },
    {
      name: 'forwardConstant',
      constant: true,
      parameters: [
        { name: 'address', type: 'Address' },
        { name: 'args', type: 'ForwardValue', rest: true },
      ],
      returnType: { type: 'ForwardValue' },
      offset: 0,
    },
    {
      name: 'forwardForward',
      parameters: [
        { name: 'address', type: 'Address' },
        { name: 'args', type: 'ForwardValue', rest: true, forwardedValue: true },
      ],
      returnType: { type: 'ForwardValue', forwardedValue: true },
      offset: 0,
    },
    {
      name: 'forwardForwardConstant',
      constant: true,
      parameters: [
        { name: 'address', type: 'Address' },
        { name: 'args', type: 'ForwardValue', rest: true, forwardedValue: true },
      ],
      returnType: { type: 'ForwardValue', forwardedValue: true },
      offset: 0,
    },
    {
      name: 'forwardTo',
      parameters: [
        { name: 'first', type: 'Integer', decimals: 8, forwardedValue: true },
        { name: 'second', type: 'Integer', decimals: 0, forwardedValue: true },
      ],
      returnType: { type: 'Integer', decimals: 8, forwardedValue: true },
      offset: 0,
    },
    {
      name: 'forwardToConstant',
      constant: true,
      parameters: [
        { name: 'first', type: 'Integer', decimals: 8, forwardedValue: true },
        { name: 'second', type: 'Integer', decimals: 0, forwardedValue: true },
      ],
      returnType: { type: 'Integer', decimals: 8, forwardedValue: true },
      offset: 0,
    },
    {
      name: 'obj',
      parameters: [
        {
          type: 'Object',
          name: 'bar',
          properties: {
            baz: {
              type: 'Object',
              properties: {
                qux: { type: 'String' },
              },
            },
          },
        },
      ],
      returnType: { type: 'Void' },
      offset: 0,
    },
  ],
  events: [
    {
      name: 'transfer',
      parameters: [
        {
          type: 'Address',
          name: 'from',
          optional: true,
        },
        {
          type: 'Address',
          name: 'to',
          optional: true,
        },
        {
          type: 'Integer',
          name: 'amount',
          decimals,
        },
      ],
    },
  ],
});

export const nep17 = {
  abi,
};
