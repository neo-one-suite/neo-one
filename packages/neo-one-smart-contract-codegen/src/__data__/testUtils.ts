import { ContractManifestClient } from '@neo-one/client-common';
import { genSmartContractTypes } from '../types';

const testContractManifestClient = (manifest: ContractManifestClient, name: string) => {
  test(name, () => {
    expect({
      inputManifest: manifest,
      types: genSmartContractTypes(name, manifest),
    }).toMatchSnapshot();
  });
};

export const testUtils = {
  testContractManifestClient,
};
