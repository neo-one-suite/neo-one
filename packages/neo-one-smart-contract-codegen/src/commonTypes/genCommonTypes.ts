import _ from 'lodash';
import { ContractPaths } from '../type';
import { getSmartContractName } from '../types';
import { getRelativeImport, lowerCaseFirst } from '../utils';

export const genCommonTypes = ({
  contractsPaths,
  commonTypesPath,
}: {
  readonly contractsPaths: ReadonlyArray<ContractPaths>;
  readonly commonTypesPath: string;
}): string => {
  const sortedPaths = _.sortBy(contractsPaths, ({ name }) => name);

  return `
${sortedPaths
    .map(
      ({ name, typesPath }) =>
        `import { ${getSmartContractName(name)} } from '${getRelativeImport(commonTypesPath, typesPath)}'`,
    )
    .join('\n')}

export interface Contracts {
  ${sortedPaths.map(({ name }) => `readonly ${lowerCaseFirst(name)}: ${getSmartContractName(name)};`).join('\n  ')}
}
`;
};
