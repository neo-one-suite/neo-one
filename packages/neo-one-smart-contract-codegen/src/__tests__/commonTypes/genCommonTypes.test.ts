import { contractsPaths } from '../../__data__/contractsPaths';
import { genCommonTypes } from '../../commonTypes';

describe('genCommonTypes', () => {
  test('Token', () => {
    expect(
      genCommonTypes({
        contractsPaths,
        commonTypesPath: '/foo/bar/one/generated/types.ts',
      }),
    ).toMatchSnapshot();
  });
});
