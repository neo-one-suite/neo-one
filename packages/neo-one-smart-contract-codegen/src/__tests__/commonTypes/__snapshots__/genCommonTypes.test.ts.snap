// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genCommonTypes Token 1`] = `
"
import { ICOSmartContract } from './ICO/types'
import { TokenSmartContract } from './Token/types'

export interface Contracts {
  readonly ico: ICOSmartContract;
  readonly token: TokenSmartContract;
}
"
`;
